### ISSUE TITLE
RequireJS and Oc-client don't really go together

### ISSUE THREAD

**Issue by navamgupta** [opentable/oc#465]
This is an interesting problem we ran into while integrating components on a different stack which uses requirejs. There are two different kind of errors we see - 

1. When clientJS is script is included after the requireJS (which is bundled with other stuff) we see the following.
```
Uncaught Error: Mismatched anonymous define() module: function (exports, $) {
      factory((root.oc = exports), $, root.head, root.document, root.window);
    }
```
Referring to - https://github.com/opentable/oc/blob/master/src/components/oc-client/_package/src/oc-client.js#L10

1. When ClientJS is included before RequireJS but the tag is included later or using js to render a component, we see the following problem - 
```
Error rendering component: http://registryurl/tutor-list/1.0.4?subject_id=43&limit=5&actions.new_tab=1&__oc_Retry=0, error: TypeError: Cannot read property 'template' of undefined
```
Basically referring to https://github.com/opentable/oc/blob/master/src/components/oc-client/_package/src/oc-client.js#L350

Thoughts? 
The 2nd problem seems like something we should be able to fix with the way we use handlebars when define is present ?

**Comment by navamgupta** [comment]
Code samples - 
1. Works
```
<!DOCTYPE html><html><head>
  <meta charset="utf-8"></head>

  <body>
    <script src="http://components-test-cdn.cheggnet.com/components/oc-client/0.34.5/src/oc-client.min.js"></script>

    <!--<script src="http://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.3/require.js"></script>-->
    <oc-component href="http://registry.chegg.com/fake-component/1.*"></oc-component>

  </body></html>

```
2. Does not work
```
<!DOCTYPE html><html><head>
  <meta charset="utf-8"></head>

  <body>
    <script src="http://components-test-cdn.cheggnet.com/components/oc-client/0.34.5/src/oc-client.min.js"></script>

    <script src="http://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.3/require.js"></script>
    <oc-component href="http://registry.chegg.com/fake-component/1.*"></oc-component>
  </body></html>

```
3. Does not work
```
<!DOCTYPE html><html><head>
  <meta charset="utf-8"></head>

  <body>

    <script src="http://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.3/require.js"></script>
    <oc-component href="http://registry.chegg.com/fake-component/1.*"></oc-component>
    <script src="http://components-test-cdn.cheggnet.com/components/oc-client/0.34.5/src/oc-client.min.js"></script>

  </body></html>

```

**Comment by matteofigus** [comment]
I think I may need a bit of @nickbalestra or @mattiaerre help here as I am not super expert about the require.js syntax.

I remember we did some work for having an AMD like syntax but that came with no tests so I am not actually sure this ever worked https://github.com/opentable/oc/pull/247

In any case, my 2 cents:
1) I don't think I am super concerned about the first problem. If we establish that OC, in case of require.js presence, has to go after that and we make it clear in the docs, I think that's acceptable?
2) In regards of the second problem, I am not sure I understand the relationship between require.js and the OC library. Handlebars should be loaded using `oc.require` and that shouldn't depend on require.js. Is it that perhaps Handlebars is already in that page, but with a version OC doesn't like?

**Comment by navamgupta** [comment]
for #1 will wait on what others have to say. 
for #2 I think the problem is 
```
if(typeof exports === 'object' && typeof module === 'object')
		module.exports = factory();
	else if(typeof define === 'function' && define.amd) 
		define([], factory);    // ends up going here
	else if(typeof exports === 'object')
		exports["Handlebars"] = factory();
	else
		root["Handlebars"] = factory(); // instead of here
```
Snippet is from the non minified version of the handlebars we load for oc -
 https://cdnjs.cloudflare.com/ajax/libs/handlebars.js/4.0.6/handlebars.runtime.js
I think handlebars wants to support amd all the way, in a sense that if amd is defined handlebars will not be exposed globally and our problem is we expect it to be [available globally](https://github.com/opentable/oc/blob/master/src/components/oc-client/_package/src/oc-client.js#L350)

**Comment by matteofigus** [comment]
I think you're right. We should somehow detect we are on AMD mode in oc require, and in case look for it via root["Handlebars"] instead of `window.Handlebars`

**Comment by mattiaerre** [comment]
I'm trying to reproduce the issue described above, with the following PR:

https://github.com/mattiaerre/oc-hub/pull/10

// cc @matteofigus @navamgupta

**Comment by navamgupta** [comment]
@mattiaerre  I tried a bunch of different scenario's last night and narrowed it down to the one scenario. 
Note: I was focused primarily on server side rendering and assuming require is loaded before the client.js
Different scenario's I played with - 
    -  With and without the markup being wrapped in <oc-component> tag even when server side rendered.
    - Using oc.require inside oc.cmd.push, something like

    <script>
        window.oc = window.oc || {};
        oc.cmd = oc.cmd || [];
        oc.cmd.push(function(oc){     // In some cases this function never executes
            oc.require("xyz.js", function(){});
           }
        });
    </script>

TL;DR : In the requireJS world if we load the client.js in the body and if we use oc.cmd.push as a method to handle oc dependency within the component it does not work. As commented on the snippet above the function never executes and drilling down further the problem seems to arise from ``` root.oc = exports ``` in [here](https://github.com/opentable/oc/blob/master/src/components/oc-client/src/oc-client.js#L11).
 As it overrides  the ```oc``` value to ```{}``` causing the ```oc.cmd``` we just did to no longer be ```undefined``` there by removing our function.

- If the client is loaded in the head : The component always renders correctly and works in all the scenario's mentioned above

- If the client is loaded in the body :  Things get a bit tricky and the scenario that breaks here is when we try to use oc.cmd.push inside the components. There are again two cases here - 
    - Client is loaded after the ssr markup : This does not work at all.
    - Client is loaded before the ssr markup : This is more complicated, its a race condition since requirejs loads client.js async, so if -
        - If the client.js completes download before oc.cmd.push line is executed, it works. 
        - If the client.js completes download after oc.cmd.push line is executed, it doesnt work.

It boils down to to the tl;dr above.

**Comment by mattiaerre** [comment]
hi @navamgupta I see your point here but, I think, you should try more to follow the requirejs approach of doing something after you load the libraries that you need; in the [PR](https://github.com/mattiaerre/oc-hub/pull/10) I created for us to reproduce the issue, you can see that it is possible to embed requirejs at the end of the body of your html and still be able to pull oc-client and then use either .cmd or .require; in particular:

```pug
html
  head
  body
    h1= title
    div!= timeago
    div!= react
    oc-component(href="https://pink-pineapple.herokuapp.com/oc-social-media-buttons/")
    script(data-main="javascripts/main" src="javascripts/require.js")
```

and 

```javascript
requirejs.config({
  paths: {
    jquery: 'https://code.jquery.com/jquery-3.2.1.min',
    'oc-client': 'http://components-test-cdn.cheggnet.com/components/oc-client/0.36.27/src/oc-client.min'
  }
});

requirejs(['jquery', 'oc-client'], ($, ocClient) => {
  console.log('ocClient:', ocClient);
  console.log('ocClient.renderedComponents:', ocClient.renderedComponents);
  console.log('ocClient.cmd:', ocClient.cmd);
  console.log('ocClient.require:', ocClient.require);
});
```

in fact w/ this approach you are sure that ocClient will be available for you to use when the callback is fired by requirejs; I think this is the right approach when using oc-client in a requirejs AMD context; what do you think?

**Comment by navamgupta** [comment]
I think how to use oc with require is exactly as you mentioned. However what I was trying to say above is that, it does not work if you add the following code 
```
<script>
    window.oc = window.oc || {};
    oc.cmd = oc.cmd || [];
    oc.cmd.push(function(){   
        alert('I am getting executed'); // See if this alert pop's up for you. 
       }
    });
</script>
```
right after the ``` oc-component(href="https://pink-pineapple.herokuapp.com/... ```  and above the ```script(data-main="javascripts/main".. ``` in your sample.
Let me know if you see the alert because I do not see it when I execute it the same way.

**Comment by mattiaerre** [comment]
@navamgupta yes you are right the alert doesn't fire; my point here is related to relying on requirejs to be sure that oc-client is ready for you to be used; the code that doesn't work is a kind of mix and match of different styles I guess; what do you think? is it acceptable for you to use the approach I am proposing or you think we should make an effort and try to see if we can improve the AMD behavior of the oc-client?

**Comment by navamgupta** [comment]
@mattiaerre The code that does not work is not really a mix and match of different styles. 
Its just a combination of - 
- Having a component which uses oc.cmd.push to wrap its js code.
- Server side rendering this component
- Loading the client.js (using proper require ways as you have explained above) after the above ssr markup. 
The script tag that I asked you to try out could be the SSR markup coming from a component right?
Maybe I am completely missing some point here, so I guess I should ask what according to you is the mix and match of styles happening here? 
If it is the fact that code coming from the component (the script i asked you to insert) itself is not using require then my argument would be that it does not make sense to re-write a component to use require just because one of the consumers (3rd party)  uses require.  Frankly I think consumer side js frameworks should not impact individual component code. The way you include the clientJS may change but the component's code should not.

On a side note, while we discuss the above I also want to understand why we have ```root.oc = exports``` in this line - https://github.com/opentable/oc/blob/master/src/components/oc-client/src/oc-client.js#L11
Is it an AMD/RequireJS standard ? If yes what does it really do?

**Comment by mattiaerre** [comment]
@navamgupta what about this as a possible solution?

https://github.com/mattiaerre/oc-hub/pull/10/files#diff-6e8c5c408e77c3d6b1af80d121354c87R11

the line you mentioned enables requirejs to have an instance of the ocClient available as a parameter of the main callback (so you can use that instead of the global `oc` object)

**Comment by matteofigus** [comment]
@mattiaerre do you think that line could change to something like `root.oc = $.extend(root.oc, exports);`?
That would preserve the oc.cmd and oc.conf in case they are pre-set before the client to be initialised

**Comment by mattiaerre** [comment]
@matteofigus I think this is not a bad idea. Let me try and I'll let you know. 

// cc @navamgupta

**Comment by mattiaerre** [comment]
It looks like we may have a good solution:

```javascript
define(['exports', 'jquery'], function(exports, $) {
  root.oc = $.extend(exports, root.oc);
  factory(root.oc, $, root.head, root.document, root.window);
});
```

what do you think @navamgupta and @matteofigus ?

**Comment by mattiaerre** [comment]
@matteofigus I am going to create a PR for this as I think it is a good starting point. thank you very much @navamgupta for your help; this has been really a joint effort 🙏

**Comment by navamgupta** [comment]
> @matteofigus I am going to create a PR for this as I think it is a good starting point. thank you very much @navamgupta for your help; this has been really a joint effort 🙏

No No, no problem at all and thanks a lot for spending a good amount of time on this with me :) Not to mention a great crash course on requireJS and AMD.

**Comment by mattiaerre** [comment]
PR's ready to be reviewed:

https://github.com/opentable/oc/pull/481
